### PR TITLE
 OutOfMemoryError when using toReceiveChannel on findBatchWithOptions or aggregate

### PR DESCRIPTION
**Pull request by bfreuden**
Motivation:

This is a fix tentative of issue #286
To be honest, it should really be scrutinized by a Vert.x savvy person because those `ReadStream`, `Publisher`, `InboundBuffer`, `Channel` interactions are definitely beyond my grasp.

Unit tests are still passing, my program no longer fails with `OutOfMemoryError`, but still... at the end of my program, the capacity of the `InboundBuffer` (792)  is slightly above the specified `batchSize` (598). It is probably not a good omen.

I also noticed this part of `MongoClientImpl`:

```java
  @Override
  public ReadStream<JsonObject> findBatchWithOptions(String collection, JsonObject query, FindOptions options) {
    requireNonNull(collection, COLLECTION_CANNOT_BE_NULL);
    requireNonNull(query, QUERY_CANNOT_BE_NULL);
    FindPublisher<JsonObject> view = doFind(collection, query, options);
    return new PublisherAdapter<>(vertx.getOrCreateContext(), view, options.getBatchSize());
  }
```
Maybe we should call the `batchSize(int)` method of the `FindPublisher`  between the `doFind` and before the `return new`?

### CODE REVIEW

File: src/main/java/io/vertx/ext/mongo/impl/PublisherAdapter.java [hunk 1]
```diff
@@ -37,6 +37,7 @@ private enum State {
   private final int batchSize;
 
   private State state;
+  private int remainingItemsInPublisherBatch;
```
Comment (tsegismont) [inline src/main/java/io/vertx/ext/mongo/impl/PublisherAdapter.java:40]: For clarity, I'm going to rename this to `requestedNotReceived` and `inFlight` to `receivedNotDelivered` before merging.

File: src/main/java/io/vertx/ext/mongo/impl/PublisherAdapter.java [hunk 2]
```diff
@@ -150,16 +152,20 @@ private void handleOut(T item) {
...
       }
       s = state;
+      remaining = remainingItemsInPublisherBatch;
+      size = internalQueue.size();
     }
     if (s == State.EXHAUSTED) {
       stop();
       handleEnd();
-    } else {
+    } else if (remaining == 0 && size == 0) {
```
Comment (tsegismont) [inline src/main/java/io/vertx/ext/mongo/impl/PublisherAdapter.java:168]: I think it's better not to wait for the queue to be empty before requesting more items from the publisher. Then Mongo cant start fetching more while we're processing the remaining items.

**Review by tsegismont — APPROVED** [review]
Thank you @bfreuden , this looks good to me. Nice catch!

I'll proceed to the changes I mentioned in comments.

**Comment by bfreuden** [comment]
@tsegismont that's perfect! Thank you :+1:

**Pull request merged**